timemachine/ff/handlers/nonbonded.py

```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
+        pos = rdmol.GetConformer().GetAtomPosition(atom_index)
+        xyz.append(f"{atom.GetSymbol()} {pos.x} {pos.y} {pos.z}")
+    return "\n".join(xyz)
+
+
+def resp_assign_partial_charges(_rdmol: Chem.Mol, use_conformers: list[Quantity]) -> tuple[np.array, float]:
+    """
+    Calculate RESP (Restrained ElectroStatic Potential) partial charges for a molecule.
+
+    This function performs quantum mechanical calculations to derive atomic partial charges
+    that best reproduce the molecular electrostatic potential while applying restraints
+    to prevent overfitting.
+
+    Parameters
+    ----------
+    _rdmol : Chem.Mol
+        Input RDKit molecule object
+    use_conformers : list[Quantity]
+        List of conformer coordinates to use (only the first conformer is used)
+
+    Returns
+    -------
+    tuple[np.array, float]
+        - Array of RESP partial charges for each atom
+        - Total DFT energy of the molecule
+    """
+    from gpu4pyscf.pop import esp
+
+    # Check that antechamber is available for symmetry checking
+    ANTECHAMBER_PATH = which("antechamber")
+    if ANTECHAMBER_PATH is None:
+        raise AntechamberNotFoundError("Antechamber not found, cannot run assign_partial_charges()")
+
+    # Create a copy of the molecule and set conformer positions
+    rdmol = Chem.Mol(_rdmol)
+    rdmol.GetConformer().SetPositions(np.array(use_conformers[0], dtype=np.float64))
+
+    # Compute charges
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # Get the formal charge of the molecule for QM calculations
+        net_charge = Chem.GetFormalCharge(rdmol)
+
+        # Write molecule to SDF file for antechamber processing
+        print(Chem.MolToMolBlock(rdmol), file=open(os.path.join(tmpdir, "molecule.sdf"), "w+"))
+
+        # Run antechamber to generate atom types and connectivity information
+        # This creates a .ac file with atom type assignments needed for RESP
+        subprocess.check_output(
+            [
+                "antechamber",
+                "-i",
+                "molecule.sdf",
+                "-o",
+                "charged.ac",
+                "-fo",
+                "ac",
+                "-fi",
+                "sdf",
+                "-nc",
+                str(net_charge),
+            ],
+            cwd=tmpdir,
+        )
+
+        # Attempt to optimize geometry using AIMNET neural network potential
+        # This can provide better initial coordinates for QM calculations
+        try:
+            out_path = opt_geometry(os.path.join(tmpdir, "molecule.sdf"), model_name="AIMNET")
+            m = Chem.MolFromMolFile(out_path, removeHs=False)
+            bad_bond = False
+            # Check for unreasonable bond lengths that indicate optimization failure
+            for bnd in m.GetBonds():
+                bond_length = GetBondLength(m.GetConformer(), bnd.GetBeginAtomIdx(), bnd.GetEndAtomIdx())
+                if bond_length > 3.0:
```

Comment:
Might be better to have the threshold be % based as opposed to a hard cutoff - i.e. if the bond lengths change +- 15% that would be quite significant.